server/app/api.py

```diff
@@ -1987,6 +2010,72 @@ def new_entity(self, attributes):
         ent.assigned_staff = [models.User.get_or_insert(
             user.id()).key for user in ent.assigned_staff]
         return ent
+    
+    
+class QueuesAPI(APIResource):
+    """ API resource for sets of queues """
+    
+    contains_entities = False
+
+    methods = {
+        'generate': {
+            'methods': set(['POST']),
+            'web_args': {
+                'course': KeyArg('Course', required=True),
+                'assignment': KeyArg('Assignment', required=True)
+            }
+        }
+    }
+    
+    def generate(self, user, data):
+        """ Splits up submissions among staff members """
+        
+        self.check_permissions(user, data)
+        course, assignment = data['course'].get(), data['assignment'].get()
+        userify = lambda parts: [part.user.get() for part in parts]
+        
+        staff = userify(models.Participant.query(
+            models.Participant.role == STAFF_ROLE,
+            models.Participant.course == course.key).fetch())
+
+        students = userify(models.Participant.query(
+            models.Participant.role == STUDENT_ROLE,
+            models.Participant.course == course.key).fetch())
+
+        subms, queues = [], []
+
+        for student in students:
+            subm = models.User.get_final_submission(student, assignment.key)
+            if subm:
+                subms.append(subm)
+        
+        for instr in staff:
+            q = models.Queue.query(
+                models.Queue.owner==instr.key,
+                models.Queue.assignment==assignment.key).get()
+            if not q:
+                q = models.Queue(
+                    owner=instr.key, 
+                    assignment=assignment.key,
+                    assigned_staff=[instr.key])
+                q.put()
+            queues.append(q)
+
+        i = 0
+
+        for subm in subms:
+            subm.queue = queues[i].key
+            subm.put()
+            i = (i + 1) % len(staff)
+            
+        return queues
+
+    def check_permissions(self, user, data):
+        course = data['course'].get()
+
+        if user.key not in course.staff and not user.is_admin:
```

Comment:
This method should probably return True/False and leave it up to the caller what to do.